devito/at_controller.py

```diff
@@ -0,0 +1,211 @@
+        self.op.propagator.profile = True
+
+        default_blocked_dims = ([True] * (len(self.op.shape) - 1))
+        default_blocked_dims.append(False)  # By default we dont auto tune inner most dim
+        self.blocked_dims = blocked_dims or default_blocked_dims
+
+        default_at_dir = path.join(path.dirname(path.realpath(__file__)), "At Report")
+        self.report_dir = None
+        self.final_report_path = None
+        self.set_report_path(at_report_dir or default_at_dir)
+        self.model_desc_template = "%s %d %d %s %s"
+
+        if not path.isdir(self.report_dir):  # Creates report dir if does not exist
+            mkdir(self.report_dir)
+
+    def set_report_path(self, report_path):
+        self.report_dir = report_path
+        self.final_report_path = path.join(self.report_dir, "final_report.txt")
+
+    def get_at_block_size(self):
+        """ Gets block size from auto tuning report
+
+        :returns: auto tuned block size
+        :raises ValueError: if auto tuning report not found
+        :raises EnvironmentError: if matching model for auto tuned block size not found
+        """
+        if not path.isfile(self.final_report_path):
+            raise ValueError("AT report at %s not found" % self.final_report_path)
+
+        # model description string
+        model_desc_str = self.model_desc_template % (self.op.getName(),
+                                                     self.op.time_order,
+                                                     self.op.spc_border,
+                                                     str(self.op.shape).replace(" ", ''),
+                                                     str(self.blocked_dims).replace(" ", ''))
+        with open(self.final_report_path, 'r') as f:
+            for line in f.readlines():
+
+                if model_desc_str in line:
+                    blocks_str = line.split(' ')[5]
+                    block_split = blocks_str[1:len(blocks_str) - 2].split(',')
+
+                    return [int(block) if block != "None" else None
+                            for block in block_split]
+
+        raise EnvironmentError("Matching model with auto tuned block size not found.")
+
+    def auto_tune_blocks(self, minimum=5, maximum=20):
+        """Auto tunes block sizes. Times all block size combinations withing given range
+           and writes it into report
+
+        :param minimum: int (optional) - minimum value for auto tuning range. Default 5
+        :param maximum: int (optional) - maximum value for auto tuning range. Default 20
+        :raises ValueError: if  minimum is >= maximum
+        """
+        if minimum >= maximum:
+            raise ValueError("Invalid parameters. Min tune range has to be less than Max")
+
+        self.op.propagator.cache_blocking = [0 if dim else None
+                                             for dim in self.blocked_dims]
+        self.op.apply()  # want to run function at least once to set the arguments
```

Comment:
Comment was not clear enough. We need to run this at least once so that block size arg when explicitly set woud not be owerwritten. As it is set based on cache_blocking argument
Im gonna change into propagator.run.